File: 200_num_islands.js
var numIslands = function (grid) {
  let numGrid = [];
  for (let row of grid){
    let newRow = [];
    for (let element of row){
      newRow.push(+element);
    }
    numGrid.push(newRow);
  }

  let islandCount = 0;
  let visited = new Set();
  for (let y = 0; y < numGrid.length; y++) {
    for (let x = 0; x < numGrid[y].length; x++) {
      if (visited.has(`${y}-${x}`) || !numGrid[y][x]) {
        continue;
      } else {
        let stack = [[y, x]];
        while (stack.length) {
          let currNode = stack.pop();
          let row = currNode[0];
          let col = currNode[1];
          visited.add(`${row}-${col}`);
          //push all adjacent 1s into stack
          if (row < numGrid.length - 1 && numGrid[row + 1][col] && !visited.has(`${row + 1}-${col}`)) {
            stack.push([row + 1, col])
          }
          if (row > 0 && numGrid[row - 1][col] && !visited.has(`${row - 1}-${col}`)) {
            stack.push([row - 1, col])
          }
          if (col < numGrid[row].length - 1 && numGrid[row][col + 1] && !visited.has(`${row}-${col + 1}`)) {
            stack.push([row, col + 1])
          }
          if (col > 0 && numGrid[row][col - 1] && !visited.has(`${row}-${col - 1}`)) {
            stack.push([row, col - 1])
          }
        }
        islandCount++;
      }
    }
  }
  return islandCount;
};

File: 200_num_islands.py
class Solution:
	def numIslands(self, grid: List[List[str]]) -> int:
		num_grid = []
		for row in grid:
			new_row = []
			for element in row:
				new_row.append(int(element))
			num_grid.append(new_row)

		island_count = 0
		visited = set()
        # iterate over every square in the grid
		for y in range(len(num_grid)):
			for x in range(len(num_grid[y])):
                # skip if square has been visited already or is a 0
				if (f'{y}-{x}' in visited) or (not num_grid[y][x]):
					continue
                # if square is part of an island that hasn't been visited,
                # dfs to add all squares of that island to visited, then increment island count
				else:
					stack = [[y, x]]
					while len(stack):
						curr_node = stack.pop()
						row = curr_node[0]
						col = curr_node[1]
						visited.add(f'{row}-{col}')
						# push all adjacent 1s into stack
						if (row < len(num_grid) - 1) and (num_grid[row + 1][col]) and (not f'{row + 1}-{col}' in visited):
							stack.append([row + 1, col])
						if (row > 0) and (num_grid[row - 1][col]) and (not f'{row - 1}-{col}' in visited):
							stack.append([row - 1, col])
						if (col < len(num_grid[row]) - 1) and (num_grid[row][col + 1]) and (not f'{row}-{col + 1}' in visited):
							stack.append([row, col + 1])
						if (col > 0) and (num_grid[row][col - 1]) and (not f'{row}-{col - 1}' in visited):
							stack.append([row, col - 1])
					island_count += 1
		return island_count